File: DFS/2667.py
def solve():
    # num = 7
    # map = [
    #     [0, 1, 1, 0, 1, 0, 0],
    #     [0, 1, 1, 0, 1, 0, 1],
    #     [1, 1, 1, 0, 1, 0, 1],
    #     [0, 0, 0, 0, 1, 1, 1],
    #     [0, 1, 0, 0, 0, 0, 0],
    #     [0, 1, 1, 1, 1, 1, 0],
    #     [0, 1, 1, 1, 0, 0, 0],
    # ]
    num = int(input())
    matrix = []
    for _ in range(num):
        matrix.append(list(map(int, input())))
    cnt = 0
    result = []
    visited = [[0] * num for _ in range(num)]

    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]
    def dfs(x, y):
        global cnt
        cnt += 1
        visited[x][y] = 1
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < num and 0 <= ny < num and visited[nx][ny] != 1 and matrix[nx][ny] == 1:
                dfs(nx, ny)

    for i in range(num):
        for j in range(num):
            if matrix[i][j] == 1 and visited[i][j] != 1:
                dfs(i, j)
                result.append(cnt);
                cnt = 0
    print(len(result))
    for i in sorted(result):
        print(i)
            
solve()

File: DFS/2667.js
let n = 7; // 지도 크기
let answer;
const map = [
	// 단지 정보를 담는 배열
	[0, 1, 1, 0, 1, 0, 0],
	[0, 1, 1, 0, 1, 0, 1],
	[1, 1, 1, 0, 1, 0, 1],
	[0, 0, 0, 0, 1, 1, 1],
	[0, 1, 0, 0, 0, 0, 0],
	[0, 1, 1, 1, 1, 1, 0],
	[0, 1, 1, 1, 0, 0, 0],
];
const visited = new Array(n); // 방문 정보 담는 배열
const result = [];
for (let i = 0; i < visited.length; i++) {
	visited[i] = new Array(n);
}
const position = [
	[-1, 0],
	[1, 0],
	[0, -1],
	[0, 1],
];
// console.log("visited 배열 초기화 : ", visited);

// 깊이 우선 탐색
function dfs(row, col) {
	console.log("행 : ", row + 1, " 열 : ", col + 1);
	answer++;
	visited[row][col] = true;
	// 상, 하, 좌, 우 탐색
	for (let i = 0; i < position.length; i++) {
		let newRow = row + position[i][0];
		let newCol = col + position[i][1];
		if (
			newRow >= 0 &&
			newRow < n &&
			newCol >= 0 &&
			newCol < n &&
			map[newRow][newCol] &&
			!visited[newRow][newCol]
		) {
			dfs(newRow, newCol);
		}
	}
}

function solution() {
	for (let i = 0; i < n; i++) {
		for (let j = 0; j < n; j++) {
			if (map[i][j] == 1 && !visited[i][j]) {
				answer = 0;
				dfs(i, j);
				result.push(answer);
			}
		}
	}
	console.log("단지 수 : ", result.length);
	console.log("결과 값 : ", result);

	return answer;
}

solution();
